### PR TITLE
feat: implement logout

### LINKED ISSUE #448: Implement Logout
We should implement the logout functionality as well.

### PR DESCRIPTION
**Pull request by Tymmmy**
### Context

- fixes #448 

### Changes
- added logout button for desktop/mobile
- redirects user to Welcome page login/signup options

### CODE REVIEW

File: packages/wallet/backend/src/auth/service.ts [hunk 1]
```diff
@@ -48,4 +48,18 @@ export class AuthService implements IAuthService {
       session
     }
   }
+
+  public async logout(id: string): Promise<void> {
+    const user = await this.deps.userService.getById(id)
+
+    // TODO: Prevent timing attacks
```
Comment (raducristianpopa) [inline packages/wallet/backend/src/auth/service.ts:55]:
```suggestion
```

File: packages/wallet/backend/src/auth/service.ts [hunk 2]
```diff
@@ -48,4 +48,18 @@ export class AuthService implements IAuthService {
...
+    if (!user) {
+      throw new Unauthorized('Invalid credentials')
+    }
+
+    await user.$relatedQuery('sessions').insertGraphAndFetch({
```
Comment (raducristianpopa) [inline packages/wallet/backend/src/auth/service.ts:60]: We want the users to only have one active session at a time?

File: packages/wallet/backend/src/auth/controller.ts [hunk 3]
```diff
@@ -63,4 +65,22 @@ export class AuthController implements IAuthController {
...
+  logOut = async (req: Request, res: CustomResponse, next: NextFunction) => {
+    try {
+      if (!req.session.id || !req.session.user) {
+        req.session.destroy()
+        throw new Unauthorized('Unauthorized')
+      }
+
+      await this.deps.authService.logout(req.session.user.id)
+      req.session.destroy()
+      res.clearCookie(env.COOKIE_NAME)
```
Comment (raducristianpopa) [inline packages/wallet/backend/src/auth/controller.ts:78]: `req.session.destroy()` automatically clears the cookie.
https://github.com/vvo/iron-session#sessiondestroy

File: packages/wallet/frontend/src/components/Menu.tsx [hunk 4]
```diff
@@ -232,6 +251,16 @@ export const Menu = () => {
               )
             )}
           </div>
+          <div className="hidden md:block">
+            <Link
```
Comment (raducristianpopa) [inline packages/wallet/frontend/src/components/Menu.tsx:255]: Considering that this is not a navigation, the component should not be a `Link`.

File: packages/wallet/frontend/src/components/Menu.tsx [hunk 5]
```diff
@@ -1,16 +1,18 @@
+import { userService } from '@/lib/api/user'
 import { MenuBubbles } from '@/ui/Bubbles'
 import { Link } from '@/ui/Link'
 import { Logo } from '@/ui/Logo'
 import { Dialog, Disclosure, Transition } from '@headlessui/react'
 import { cx } from 'class-variance-authority'
-import { useRouter } from 'next/router'
+import router, { useRouter } from 'next/router'
```
Comment (raducristianpopa) [inline packages/wallet/frontend/src/components/Menu.tsx:7]:
```suggestion
import { useRouter } from 'next/router'
```

File: packages/wallet/frontend/src/components/Menu.tsx [hunk 6]
```diff
@@ -70,6 +72,13 @@ export const Menu = () => {
   const pathname = `/${useRouter().pathname.split('/')?.slice(1)[0] ?? ''}`
   const [sidebarIsOpen, setSidebarIsOpen] = useState(false)
 
+  const handleLogout = async () => {
+    const res = await userService.logout()
+    if (res.success) {
+      router.push('/auth')
+    }
+  }
+
```
Comment (raducristianpopa) [inline packages/wallet/frontend/src/components/Menu.tsx:81]:
```suggestion
  const router = useRouter()
  const pathname = `/${pathname.split('/')?.slice(1)[0] ?? ''}`
  const [sidebarIsOpen, setSidebarIsOpen] = useState(false)

  const handleLogout = async () => {
    const res = await userService.logout()
    if (res.success) {
      router.push('/auth')
    }
  }

```

File: packages/wallet/backend/src/auth/service.ts [hunk 7]
```diff
@@ -48,4 +48,16 @@ export class AuthService implements IAuthService {
...
+  public async logout(id: string): Promise<void> {
+    const user = await this.deps.userService.getById(id)
+
+    if (!user) {
+      throw new Unauthorized('Invalid credentials')
+    }
+
+    await user.$relatedQuery('sessions').where({ userId: id }).update({
+      expiresAt: new Date()
+    })
```
Comment (raducristianpopa) [inline packages/wallet/backend/src/auth/service.ts:61]: Shouldn't this be the session ID instead of the user ID ?

**Review by beniaminmunteanu — APPROVED** [review]

**Pull request merged**